sbi/neural_nets/embedding_nets.py

```diff
@@ -44,81 +45,151 @@ def forward(self, x: Tensor) -> Tensor:
         """
-        super().__init__()
-        self.input_dim = input_dim
-        self.output_dim = output_dim
-        self.num_hiddens = num_hiddens
-
-        # construct convolutional-pooling subnet
-        pool = nn.MaxPool1d(pool_size)
-        conv_layers = [
-            nn.Conv1d(1, out_channels_cnn_1, kernel_size, padding="same"),
-            nn.ReLU(),
-            pool,
-            nn.Conv1d(
-                out_channels_cnn_1, out_channels_cnn_2, kernel_size, padding="same"
-            ),
-            nn.ReLU(),
-            pool,
-        ]
-        self.conv_subnet = nn.Sequential(*conv_layers)
+        super(CNNEmbedding2D, self).__init__()
 
-        # construct fully connected layers
-        input_dim_fc = out_channels_cnn_2 * (int(input_dim / out_channels_cnn_2))
+        assert isinstance(
+            input_shape, Tuple
+        ), "input_shape must be a Tuple of size 1 or 2, e.g., (width, [height])."
+        assert (
+            0 < len(input_shape) < 3
+        ), """input_shape must be a Tuple of size 1 or 2, e.g.,
+            (width, [height]). Number of input channels are passed separately"""
 
-        self.fc_subnet = FCEmbedding(
-            input_dim=input_dim_fc,
+        use_2d_cnn = len(input_shape) == 2
+        if use_2d_cnn:
+            conv_module = nn.Conv2d
+            pool_module = nn.MaxPool2d
+            assert (
+                input_shape[0] == input_shape[1]
+            ), """input_shape must square e.g, (32, 32)."""
+        else:
+            conv_module = nn.Conv1d
+            pool_module = nn.MaxPool1d
+            input_shape[0]
+        assert (
+            len(out_channels_per_layer) == num_conv_layers
+        ), "out_channels needs as many entries as num_cnn_layers."
+
+        # define input shape with channel
+        self.input_shape = (in_channels, *input_shape)
+
+        # Construct CNN feature extractor.
+        cnn_layers = []
+        cnn_output_size = self.input_shape[1]
+        stride = 1
+        padding = 1
+        for ii in range(num_conv_layers):
+            # Defining another 2D convolution layer
+            conv_layer = conv_module(
+                in_channels=in_channels if ii == 0 else out_channels_per_layer[ii - 1],
+                out_channels=out_channels_per_layer[ii],
+                kernel_size=kernel_size,
+                stride=stride,
+                padding=padding,
+            )
+            pool = pool_module(kernel_size=pool_kernel_size)
+            cnn_layers += [conv_layer, nn.ReLU(inplace=True), pool]
+            # Calculate change of output size of each CNN layer
+            cnn_output_size = get_new_cnn_output_size(cnn_output_size, conv_layer, pool)
+
+        self.cnn_subnet = nn.Sequential(*cnn_layers)
+
+        # Construct linear post processing net.
+        self.linear_subnet = FCEmbedding(
+            input_dim=out_channels_per_layer[-1] * cnn_output_size * cnn_output_size
```

Comment:
if `get_new_cnn_output_size` returns a tuple one could simply do `torch.prod()` here and avoid the if case